pulser/simulation/simresults.py

```diff
@@ -108,122 +402,130 @@ def get_final_state(self, reduce_to_basis: Optional[str] = None,
-            weights = []
-            for dec_val in range(2**N):
+            probs = probs.reshape([3]*n)
+            weights = np.zeros(2**n)
+            for dec_val in range(2**n):
                 ind: list[Union[int, slice]] = []
-                for v in np.binary_repr(dec_val, width=N):
+                for v in np.binary_repr(dec_val, width=n):
                     if v == '0':
                         ind.append(ex_one)
                     else:
                         ind.append(one_state)
-                # Eg: 'digital' basis => |1> = index 2, |0> = index 0, 1 = 0:2
+                # Eg: 'digital' basis : |1> = index2, |0> = index0, 1 = 0:2
                 # p_11010 = sum(probs[2, 2, 0:2, 2, 0:2])
-                # We sum all probabilites that correspond to measuring 11010,
-                # namely hhghg, hhrhg, hhghr, hhrhr
-                weights.append(np.sum(probs[tuple(ind)]))
+                # We sum all probabilites that correspond to measuring
+                # 11010, namely hhghg, hhrhg, hhghr, hhrhr
+                weights[dec_val] = np.sum(probs[tuple(ind)])
         else:
             raise NotImplementedError(
                 "Cannot sample system with single-atom state vectors of "
-                "dimension > 3."
-                )
-        dist = np.random.multinomial(N_samples, weights)
-        return {np.binary_repr(i, N): dist[i] for i in np.nonzero(dist)[0]}
+                "dimension > 3.")
+        # Takes care of numerical artefacts in case sum(weights) != 1
+        weights /= sum(weights)
+        return cast(np.ndarray, weights)
+
+    def _sampling_with_detection_errors(self, spam: dict[str, float],
+                                        t: float,
+                                        n_samples: int = 1000) -> Counter:
+        """Returns the distribution of states really detected.
+
+        Part of the SPAM implementation.
+
+        Args:
+            spam (dict): Dictionnary gathering the SPAM error
+            probabilities.
+            t (float): Time at which to return the samples.
+            n_samples (int): Number of samples.
+        """
+        def detection_from_basis_state(n_detects: int, shot: str) -> Counter:
+            """Returns distribution of states detected when detecting `shot`.
+
+            Part of the SPAM implementation : computes measurement errors.
+
+            Args:
+                n_detects (int): Number of times state has been detected.
+                shot (str): Binary string of length the number of atoms of the
+                simulation.
+            """
+            detected_dict: Counter = Counter()
+            eps = spam['epsilon']
+            eps_p = spam['epsilon_prime']
+            # Probability of flipping each bit
+            flip_probs = [eps_p if x == '1' else eps for x in shot]
+            for _ in range(n_detects):
+                shots = [(np.random.uniform() < p)*1 for p in flip_probs]
```

Comment:
You should take advantage of numpy here:
```suggestion
                shots = (np.random.uniform(size=len(flip_probs)) < flip_probs).astype(int)
```

Reply:
Yes, I tried that, but it's actually way slower...